HPCC-17851 Phase 1 new configuration manager - core library

kenrowland opened the pull request:
Add core library for reading configuration XSDs and environment.
Provide classes to read environment and set values.

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

Commit:
HPCC-17851 Phase 1 new configuration manager - core library

Add core library for reading configuration XSDs and environment.
Provide classes to read environment and set values.

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

Commit:
HPCC-17851 Phase 1 core library for new configuration manager

Add core library to support full XSD compliant schema definitions.
Remove custom code in favor of data driven limits and dependencies.

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

kenrowland commented:
@rpastrana @RussWhitehead @afishbeck 
Please review. There will be a meeting first to discuss class structure to give context so you can wait before looking at the code.

kenrowland commented:
@wangkx Adding

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 69cd02e4a970847eca1849c67b1f87b45dd8c8e1
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 41.742 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |92 |92 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |744 |744 |0 |
|test (thor) |639 |639 |0 |
|test (roxie) |772 |772 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

rpastrana commented on configuration/config2/CfgIntegerLimits.hpp:
```diff
@@ -0,0 +1,34 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2015 HPCC Systems®.
```
update year to file creation year.

kenrowland replied:
Done

rpastrana commented on configuration/config2/configfiles/types.xsd:
```diff
@@ -0,0 +1,65 @@
+    </xs:attributeGroup>
+	<xs:attributeGroup name="computerNodeReference">
+        <xs:attribute name="computer" type="nodeName" use="required" hpcc:readOnly="true" hpcc:mirrorFrom="/Environment/Hardware/Computer@name"/>
+        <xs:attribute name="netAddress" type="ipV4Address" use="required" hpcc:readOnly="true" hpcc:mirrorFrom="/Environment/Hardware/Computer@netAddress"/>
```
I'm not sure how relevant this is, but configmanager has to allow ip addresses and hostnames in all netaddress type fields.

kenrowland replied:
The computerNodeReference attribute group is for use by any component that needs to list computer instances to which their component is bound. It marks the nodeName and ip address as readonly so that they can only view these values for inspection only. The source, in the Hardware component, is the only place where these values can be changed. the mirrorFrom keyword ensures that when a value is changed in the Hardware component, the value gets replicated to any component using this attribute group.

rpastrana replied:
ok so are you allowing hostnames (not to be confused with the name used to reference a computer node within confmgr)? It doesn't look like we do, please make sure we do.

rpastrana commented on configuration/config2/configfiles/types.xsd:
```diff
@@ -0,0 +1,65 @@
+        </xs:restriction>
+    </xs:simpleType>
+	<xs:simpleType name="absolutePath">
+		<xs:restriction base="xs:string">
```
Ensure we support POSIX standards.

rpastrana replied:
Did you ensure POSIX standards?

rpastrana commented on configuration/config2/configfiles/newenv.xsd:
```diff
@@ -0,0 +1,44 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Does this XSD represent the overall environment configuration definition? If so, the filename should reflect that.

kenrowland replied:
We discussed. This is for testing the layout of the overall environment.xsd. I'll start building that one soon.

rpastrana commented on configuration/config2/configfiles/newenv.xsd:
```diff
@@ -0,0 +1,44 @@
+        <xs:element name="EnvSettings" hpcc:class="category" hpcc:displayName="System Settings" hpcc:hidden="true"></xs:element>
+        <xs:element type="hardware" hpcc:class="category" hpcc:displayName="Hardware" hpcc:required="true"></xs:element>
+        <xs:element name="Programs" hpcc:class="category" hpcc:displayName="Programs" hpcc:hidden="true"></xs:element>
+        <xs:element name="EXTRA_SECTION" hpcc:class="category" hpcc:displayName="More Stuff" hpcc:required="false"></xs:element>
```
Did this line sneak in?

rpastrana commented on configuration/config2/configfiles/hardware.xsd:
```diff
@@ -0,0 +1,108 @@
+    <xs:sequence>
+      <xs:element name="Hardware" hpcc:docid="????">
+        <xs:annotation>
+          <xs:documentation>Section to describe the hardware that composes the environment.</xs:documentation>
```
Have @JamesDeFabia look over the messages.

kenrowland replied:
My plan was for each component owner to go over the messages, tooltips, and any other documentation items. The owners should work with the doc team to get the correct wording. In addition, the doc team should take translations into account by collecting all of the types of locations where translatable text lives. For example, all hpcc:tooltip="text" and hpcc:displayName="text" occurrences.

rpastrana commented on configuration/config2/XSDValueSetParser.hpp:
```diff
@@ -0,0 +1,56 @@
+        XSDValueSetParser(const std::string &basePath, std::shared_ptr<ConfigItem> pValueSet) : 
+            XSDConfigParser(basePath, pValueSet) { m_pValueSet = std::dynamic_pointer_cast<ConfigItemValueSet>(pValueSet);  }
+        virtual ~XSDValueSetParser() { }
+        virtual void parseXSD(const pt::ptree &valueSetTree);
```
If any of these are overrides, go ahead and declare them as such.

kenrowland replied:
Just wondering if this is a standard we are trying to follow. I like using the keyword, but wasn't sure if it was used by the team. I will begin adding it as I move through the code.

rpastrana replied:
Yes, we're all encouraged to pick this up as a standard.

rpastrana commented on configuration/config2/XSDValueSetParser.hpp:
```diff
@@ -0,0 +1,56 @@
+
+};
+
+
```
you'll want to remove all superfluous new lines

kenrowland replied:
OK. I figured there would be a massive cleanup before the final pull request anyway.

rpastrana commented on configuration/config2/XSDValueSetParser.hpp:
```diff
@@ -0,0 +1,56 @@
+#include "XSDConfigParser.hpp"
+#include "ConfigItemValueSet.hpp"
+
+namespace pt = boost::property_tree;
```
was there a reason to choose boost::property_tree over a jlib equivalent?

kenrowland replied:
Much of it was because of prior knowledge of how to use boost property trees. Looking at the jlib trees didn't show much in the way of documentation and looking at code examples showed that it would have taken longer to get the code up and running. We can discuss later if you like.

afishbeck replied:
The thing to ask ourselves is whether this is going to make others reluctant to make any changes to this code as everyone is more familiar with JLIB.

rpastrana commented on configuration/config2/XSDConfigParser.cpp:
```diff
@@ -0,0 +1,469 @@
+    bool rc = true;
+    try
+    {
+		//
```
It looks like tabs might have snuck in, make sure to use white spaces instead.

rpastrana commented on configuration/config2/XSDConfigParser.cpp:
```diff
@@ -0,0 +1,469 @@
+    {
+		//
+		// Add some default types to the config. Note changing values for limits 
+		std::shared_ptr<CfgLimits> pStringLimits = std::make_shared<CfgStringLimits>();;
```
extra ";"

rpastrana commented on configuration/config2/ConfigParser.cpp:
```diff
@@ -0,0 +1,39 @@
+}
+
+
+std::vector<std::string> ConfigParser::split(const std::string  &input, const std::string  &delim)
```
this looks to be static in nature (no member vars are referenced). In fact it's generic enough to be a util type of function. So, think about making it static, think about abstracting it out to a general use location and see if jLib already provides this function (for StringBuffers obviously).

rpastrana commented on configuration/config2/EnvValue.cpp:
```diff
@@ -0,0 +1,114 @@
+bool EnvValue::setValue(const std::string &value, bool force) 
+{ 
+	bool rc = true;
+    std::string oldValue = m_value;
```
ok so m_value is THE value, and m_pCfgValue is the definition for that value right?
Idk, CfgValue might be good enough, but perhaps a slight name change will be more informative
how about p_valueDefinition ?

rpastrana replied:
Above comment wasn't addressed.

rpastrana commented on configuration/config2/EnvValue.cpp:
```diff
@@ -0,0 +1,114 @@
+	}
+	else
+	{
+		addStatus(warning, "no configuration defined for this value, unable to validate");
```
"no definition for this value" or something like that. the content of the XSDs isn't the configuration, but rather the configuration definition right?

rpastrana commented on configuration/config2/EnvironmentMgr.cpp:
```diff
@@ -0,0 +1,122 @@
+    {
+        //std::shared_ptr<ConfigParser> pCfgParser = std::make_shared<XSDConfigParser>(configPath, m_pConfig);
+        pEnvMgr = new XMLEnvironmentMgr(configPath);
+    }
```
what if envtype != xml?

rpastrana commented on configuration/config2/EnvironmentMgr.cpp:
```diff
@@ -0,0 +1,122 @@
+
+bool EnvironmentMgr::loadConfig(const std::vector<std::string> &cfgParms)
+{
+    createParser(cfgParms);
```
does createParser() set m_pConfigParser?

rpastrana replied:
Ensure m_pConfigParser is initialized. see question above

rpastrana commented on configuration/config2/EnvironmentMgr.cpp:
```diff
@@ -0,0 +1,122 @@
+
+bool EnvironmentMgr::loadEnvironment(const std::string &filename)
+{
+	std::ifstream in;
```
not familiar w these, does in need to be closed?

rpastrana commented on configuration/config2/EnvironmentMgr.cpp:
```diff
@@ -0,0 +1,122 @@
+
+void EnvironmentMgr::saveEnvironment(const std::string &filename)
+{
+	std::ofstream out;
```
does out need to be closed? Does it get closed automatically or by caller?

rpastrana replied:
ensure out is closed if need be. See above question.

rpastrana commented on configuration/config2/EnvironmentMgr.cpp:
```diff
@@ -0,0 +1,122 @@
+bool EnvironmentMgr::validate()
+{
+	if (m_pRootNode)
+	{
```
todo?

rpastrana commented on configuration/config2/EnvironmentMgr.cpp:
```diff
@@ -0,0 +1,122 @@
+EnvironmentMgr *getEnvironmentMgrInstance(const std::string &envType, const std::string &configPath)
+{
+    EnvironmentMgr *pEnvMgr = NULL;
+    if (envType == "XML")
```
can these types be enumerated? (I'm sure you currently only have xml, but let's set up for the future).

rpastrana commented on configuration/config2/EnvironmentMgr.hpp:
```diff
@@ -0,0 +1,81 @@
+#include "ConfigParser.hpp"
+#include "EnvironmentNode.hpp"
+
+class EnvironmentMgr;
```
sometime Environment gets abbreviated, sometimes it doesn't. is that just an inconsistency or is there a hidden meaning behind the naming scheme?

rpastrana commented on configuration/config2/EnvironmentNode.cpp:
```diff
@@ -0,0 +1,210 @@
+
+void EnvironmentNode::addAttribute(const std::string &name, std::shared_ptr<EnvValue> pValue)
+{
+	auto retValue = m_attributes.insert(std::make_pair(name, pValue));
```
do we not know the return type? when you start using retValue, it might be easier for reviewers to know the type.

rpastrana reviewed (changes requested):
@kenrowland first set of comments, more coming later.

kenrowland commented on configuration/config2/CfgStringLimits.hpp:
```diff
@@ -0,0 +1,44 @@
+
+        CfgStringLimits() : m_removeWhiteSpace(true) { };
+        virtual ~CfgStringLimits() { };
+        void setRemoveWhiteSpace(bool remove) { m_removeWhiteSpace = true; }
```
I was not quite sure where to put this. There is a string restriction to collapse (remove) all whitespace from a value. The restriction is for the XML processor (I believe when the schema is used to validate XML, this tells the processor to remove the whitespace), but since we don't have one that is used, we needed to capture it, just in case. Since the CfgType is a lightweight class and all of the actual type checking is done by the Limits object, I placed it there. While not using it yet, the intent was that when a value was checked for validity through the Limits object, if this flag was set, whitespace would be removed prior to saving the value to the environment.

<xs:restriction base="xs:string">
      <xs:whiteSpace value="collapse"/>
</xs:restriction>

kenrowland commented on configuration/config2/CfgValue.hpp:
```diff
@@ -0,0 +1,94 @@
+        void addModifer(const std::string &mod) { m_modifiers.push_back(mod); }
+        void setModifiers(const std::vector<std::string> &list) { m_modifiers = list;  }
+        const std::vector<std::string> &getModifiers() const { return m_modifiers; }
+        bool hasModifiers() const { return m_modifiers.size() > 0; }
```
size cannot be < 0, but changed to != 0;

kenrowland commented on configuration/config2/CfgValue.hpp:
```diff
@@ -0,0 +1,94 @@
+        void resetEnvironment();
+        void setMirrorFromPath(const std::string &path) { m_mirrorFromPath = path;  }
+        const std::string &getMirrorFromPath() const { return m_mirrorFromPath;  }
+        bool isMirroredValue() const { return m_mirrorFromPath.length() > 0; }
```
Changed to != 0

kenrowland commented on configuration/config2/ConfigItem.cpp:
```diff
@@ -0,0 +1,322 @@
+    // Did not find the type
+    if (throwIfNotPresent)
+    {
+        std::string msg = "Unable to find config type: " + name;
```
The exceptions will get expanded.

rpastrana commented on configuration/config2/EnvValue.cpp:
```diff
@@ -0,0 +1,114 @@
+				std::shared_ptr<EnvironmentNode> pMyEnvNode = m_pMyEnvNode.lock();
+				if (pMyEnvNode)
+				{
+                    // todo: use getAllFieldValues
```
it seems there's more todo than "use getAllFieldValues". until this is implemented it looks like we'll allow key conflicts

rpastrana commented on configuration/config2/EnvValue.cpp:
```diff
@@ -0,0 +1,114 @@
+
+bool EnvValue::validate() const
+{
+    return isValueValid(m_value);
```
is this extra layer needed?

rpastrana commented on configuration/config2/EnvironmentMgr.cpp:
```diff
@@ -0,0 +1,122 @@
+/*##############################################################################
+
+HPCC SYSTEMS software Copyright (C) 2015 HPCC Systems�.
```
odd character (c) ?

rpastrana reviewed (changes requested):
@kenrowland a couple more comments

Commit:
Update with review comments and additional function.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: d0b47e36f9bd27c73564125b51df81f5330ccd48
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |727 |727 |0 |
|test (thor) |640 |640 |0 |
|test (roxie) |755 |755 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

rpastrana commented on configuration/config2/CfgValue.cpp:
```diff
@@ -19,14 +19,48 @@
+
+void CfgValue::validate(Status &status, const std::string &id, const EnvValue *pEnvValue) const
+{
+    // get currentvalue from pEnvValue
```
is it ok to have an empty implementation for this meth?

rpastrana commented on configuration/config2/CfgValue.hpp:
```diff
@@ -20,52 +20,63 @@
+        std::vector<std::string> getAllEnvValues() const;
         void mirrorValue(const std::string &oldValue, const std::string &newValue);
         void setMirroredEnvValues(const std::string &oldValue, const std::string &newValue);
+        //void validate(Status &status, const std::string &id, const std::string &value, const EnvValue *pEnvValue = nullptr) const;
```
let's remove all commented out code

rpastrana commented on configuration/config2/CfgValue.hpp:
```diff
@@ -20,52 +20,63 @@
+        void validate(Status &status, const std::string &id, const EnvValue *pEnvValue = nullptr) const;
+        std::vector<AllowedValue> getAllowedValues(const EnvValue *pEnvValue = nullptr) const;
+
+
```
remove unneeded blank lines

rpastrana commented on configuration/config2/CfgValue.hpp:
```diff
@@ -20,52 +20,63 @@
+        std::vector<AllowedValue> getAllowedValues(const EnvValue *pEnvValue = nullptr) const;
+
+
+    protected:
```
do we need the protected label in line 82?

rpastrana commented on configuration/config2/ConfigItem.hpp:
```diff
@@ -93,22 +104,38 @@ class ConfigItem : public std::enable_shared_from_this<ConfigItem>
-        std::map<std::string, std::shared_ptr<ConfigItem>> m_children; 
-        std::shared_ptr<CfgValue> m_pValue;   // value for this item (think of it as the VALUE for an element <xx attr= att1=>VALUE</xx>)
-		std::map<std::string, std::shared_ptr<CfgValue>> m_attributes;   // attributes for this item (think in xml terms <m_name attr1="val" attr2="val" .../> where attrN is in this vector
+        std::string m_componentName;   
```
extra spaces. Your environment might be lacking the git hooks Richard created which prevents these types of issues to be allowed when you commit/push.

rpastrana commented on configuration/config2/ConfigItem.hpp:
```diff
@@ -93,22 +104,38 @@ class ConfigItem : public std::enable_shared_from_this<ConfigItem>
+
+        // This struct handles both unique attribute sets and references to same
+        struct SetInfo {
+            SetInfo(const std::string &setName, const std::string &elementPath, const std::string &attributeName) :
```
indent seems off

rpastrana commented on configuration/config2/ConfigItemComponent.hpp:
```diff
@@ -28,7 +28,7 @@ class ConfigItemComponent : public ConfigItem
     public:
 
-		ConfigItemComponent(const std::string &name, std::shared_ptr<ConfigItem> pParent) : ConfigItem(name, "component", pParent) { m_isConfigurable = true; };
+		ConfigItemComponent(const std::string &name, std::shared_ptr<ConfigItem> pParent) : ConfigItem(name, "component", pParent) { }
```
indent

rpastrana commented on configuration/config2/ConfigItemValueSet.cpp:
```diff
@@ -31,7 +31,7 @@ void ConfigItemValueSet::addCfgValue(const std::shared_ptr<CfgValue> pNewValue)
     else 
     {
         std::string msg = "A value already exists for " + pNewValue->getName();
```
Consider providing more context: 
eg "ValueSet: "this->getName()" + "Already contains a value named: " + pNewValue->getName()

rpastrana commented on configuration/config2/ConfigItemValueSet.cpp:
```diff
@@ -0,0 +1,67 @@
+    }
+    else if (throwIfNotFound)
+    {
+        std::string msg = "Unable to find valueName(" + valueName + ")";
```
Above comment wasn't addressed.

rpastrana commented on configuration/config2/ConfigParser.cpp:
```diff
@@ -16,11 +16,25 @@
+    }
+    catch (...)
+    {
+        int i = 8;
```
what is this?

rpastrana commented on configuration/config2/EnvValue.cpp:
```diff
@@ -1,6 +1,6 @@
 /*##############################################################################
 
-HPCC SYSTEMS software Copyright (C) 2015 HPCC Systems�.
+HPCC SYSTEMS software Copyright (C) 2017 HPCC Systems�.
```
odd char after "Systems"

rpastrana commented on configuration/config2/EnvValue.cpp:
```diff
@@ -16,99 +16,83 @@ limitations under the License.
+            m_forcedSet = true;
+            m_pCfgValue->mirrorValue(oldValue, value);
+            if (pStatus != nullptr)
+                pStatus->addStatusMsg(statusMsg::info, m_pMyEnvNode.lock()->getId(), m_name, "", "Attribute forced to invalid value");
```
This message isn't immediately obvious, can it be enhanced to help in debug?

rpastrana commented on configuration/config2/EnvValue.cpp:
```diff
@@ -16,99 +16,83 @@ limitations under the License.
+        else
+        {
+            if (pStatus != nullptr)
+                pStatus->addStatusMsg(statusMsg::error, m_pMyEnvNode.lock()->getId(), m_name, "", "New value is not valid");
```
what is the value that is not valid?

rpastrana commented on configuration/config2/EnvValue.cpp:
```diff
@@ -16,99 +16,83 @@ limitations under the License.
-    return isValueValid(m_value);
+
+    if (!m_pCfgValue->isConfigured())
+        status.addStatusMsg(statusMsg::warning, myId, m_name, "", "No configuration exists for this value");
```
what is the value? be as informative as possible.

rpastrana commented on configuration/config2/EnvValue.cpp:
```diff
@@ -16,99 +16,83 @@ limitations under the License.
+        status.addStatusMsg(statusMsg::warning, myId, m_name, "", "No configuration exists for this value");
+
+    if (m_forcedSet)
+        status.addStatusMsg(statusMsg::warning, myId, m_name, "", "Current value was force set");
```
what does this mean to user/person debugging

rpastrana commented on configuration/config2/EnvValue.hpp:
```diff
@@ -1,6 +1,6 @@
 /*##############################################################################
 
-HPCC SYSTEMS software Copyright (C) 2015 HPCC Systems�.
+HPCC SYSTEMS software Copyright (C) 2017 HPCC Systems�.
```
odd char after "Systems" in github.

rpastrana commented on configuration/config2/EnvValue.hpp:
```diff
@@ -20,28 +20,38 @@ limitations under the License.
+		bool wasForced() const { return m_forcedSet; }
 		bool isValueValid(const std::string &value) const;
-        bool validate() const;
+        void validate(Status &status, const std::string &myId) const;
```
indent

rpastrana commented on configuration/config2/EnvironmentMgr.cpp:
```diff
@@ -1,6 +1,6 @@
 /*##############################################################################
 
-HPCC SYSTEMS software Copyright (C) 2015 HPCC Systems�.
+HPCC SYSTEMS software Copyright (C) 2017 HPCC Systems�.
```
odd char after "Systems"

rpastrana commented on configuration/config2/EnvironmentMgr.cpp:
```diff
@@ -78,32 +84,39 @@ void EnvironmentMgr::saveEnvironment(const std::string &filename)
+	std::shared_ptr<EnvironmentNode> pNode;
+	auto pathIt = m_nodeIds.find(nodeId);
+	if (pathIt != m_nodeIds.end())
+        pNode = pathIt->second;
```
indent

rpastrana commented on configuration/config2/EnvironmentMgr.cpp:
```diff
@@ -78,32 +84,39 @@ void EnvironmentMgr::saveEnvironment(const std::string &filename)
 
 // todo: make a standard return that has a status string and array of messages
-bool EnvironmentMgr::setValuesForPath(const std::string &path, const std::vector<valueDef> &values, const std::string &nodeValue, bool force)
+/*void EnvironmentMgr::setAttributeValues(const std::string &nodeId, const std::vector<valueDef> &values, const std::string &nodeValue, bool force)
```
is this needed? remove otherwise.

rpastrana commented on configuration/config2/EnvironmentMgr.hpp:
```diff
@@ -46,18 +55,18 @@ class EnvironmentMgr
 
-		std::shared_ptr<EnvironmentNode> getNodeFromPath(const std::string &path) { return getElement(path); }
-		bool setValuesForPath(const std::string &path, const std::vector<valueDef> &values, const std::string &nodeValue, bool force=false);
+        std::shared_ptr<EnvironmentNode> getEnvironmentNode(const std::string &nodeId);
```
indenting seems inconsistent  lines 55-84

rpastrana commented on configuration/config2/EnvironmentNode.cpp:
```diff
@@ -19,36 +19,36 @@ limitations under the License.
-		}
-	}
+    std::vector<std::shared_ptr<EnvironmentNode>> childNodes;
+    if (name == "") {
```
can we condense the if and else loop into one, by setting the rangeIt appropriately?

rpastrana commented on configuration/config2/EnvironmentNode.cpp:
```diff
@@ -73,138 +73,166 @@ std::map<std::string, std::vector<std::shared_ptr<EnvironmentNode>>> Environment
 
 // should probably return a status object, and put path/valueName in there
-bool EnvironmentNode::setAttributeValue(const std::string &name, const std::string &value, bool allowInvalid, bool forceCreate)
+void EnvironmentNode::setAttributeValue(const std::string &attrName, const std::string &value, Status &status, bool allowInvalid, bool forceCreate)
```
then let's do so... "// should probably return a status object, and put path/valueName in there"

rpastrana commented on configuration/config2/XMLEnvironmentMgr.cpp:
```diff
@@ -84,16 +84,22 @@ void XMLEnvironmentMgr::parse(const pt::ptree &envTree, const std::shared_ptr<Co
 		std::string elemName = it->first;
 
+        if (elemName == "EspService")
+            int i = 3;
```
is this being used? And if so, it seems very arbitrary

rpastrana commented on configuration/config2/XMLEnvironmentMgr.cpp:
```diff
@@ -84,16 +84,22 @@ void XMLEnvironmentMgr::parse(const pt::ptree &envTree, const std::shared_ptr<Co
+				std::string curValue = attrIt->second.get_value<std::string>();
+				std::shared_ptr<EnvValue> pEnvValue = std::make_shared<EnvValue>(pEnvNode, pCfgValue, attrIt->first, curValue);   // this is where we would use a variant
+                pCfgValue->addEnvValue(pEnvValue);
+                //auto x = pCfgValue.get();
```
remove

rpastrana commented on configuration/config2/XSDComponentParser.cpp:
```diff
@@ -104,42 +114,42 @@ void XSDComponentParser::parseXSD(const pt::ptree &compTree)
-
-    m_pConfig->addKeyRef(keyRefName, elementName, attributeName);
-}
+//void XSDComponentParser::parseKey(const pt::ptree &keyTree)
```
what's the plan w/ this method?

rpastrana commented on configuration/config2/XSDComponentParser.hpp:
```diff
@@ -41,8 +41,8 @@ class XSDComponentParser : public XSDConfigParser
         XSDComponentParser() { };
-        virtual void parseKey(const pt::ptree &keyTree);
-        virtual void parseKeyRef(const pt::ptree &keyTree);
+        //virtual void parseKey(const pt::ptree &keyTree);
```
remove?

rpastrana commented on configuration/config2/cfg20.sln:
```diff
@@ -0,0 +1,31 @@
+﻿
```
no IDE specific code

afishbeck replied:
right, if you get cmake right you can generate files for VS.

rpastrana commented on configuration/config2/cfg20.vcxproj:
```diff
@@ -0,0 +1,168 @@
+<?xml version="1.0" encoding="utf-8"?>
```
remove

rpastrana commented on configuration/config2/configfiles/esp.xsd:
```diff
@@ -0,0 +1,210 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
can Ming generate docs off of this new XSD?

rpastrana reviewed (changes requested):
@kenrowland  lots of comments. Some of them have to do w/ editing, please clean-up comments, white space, and license header. Please confirm w/ @JamesDeFabia  what (c) should be used (some of your headers have "(c)" and the actual copyright char and some don't have the copyright char, some have an odd char).

afishbeck commented on configuration/config2/CfgIntegerLimits.cpp:
```diff
@@ -0,0 +1,28 @@
+bool CfgIntegerLimits::isValueValid(const std::string &value)
+{
+    bool isValid = true;
+    int testValue = std::stoi(value);
```
If you change this to throw an exception I would change the name to "checkValueIsValid()"

rpastrana replied:
I believe std::stoi(value) throws (i'm not 100% familiar w/ those calls, @kenrowland  should confirm) but if it does, CfgIntegerLimits::isValueValid does not catch it.

afishbeck commented on configuration/config2/CfgIntegerLimits.cpp:
```diff
@@ -0,0 +1,28 @@
+    isValid = testValue >= getMin() && testValue <= getMax();
+
+    return isValid;
+}
```
That symbol means this needs a newline.

rpastrana replied:
new line is still needed

afishbeck commented on configuration/config2/CfgStringLimits.cpp:
```diff
@@ -0,0 +1,75 @@
+
+    if (m_length != 0)
+    {
+        //limitStr += "string " + m_length + " charcters in length.";
```
Just a note, If this were a final code review I'd ask that commented out code be removed.

afishbeck commented on configuration/config2/CfgStringLimits.cpp:
```diff
@@ -0,0 +1,75 @@
+
+std::string CfgStringLimits::getString() const
+{
+
```
trivial, a lot of excess whitespace

afishbeck commented on configuration/config2/CfgLimits.hpp:
```diff
@@ -0,0 +1,79 @@
+
+    protected:
+
+        int m_minInclusive;
```
Is some of this wasteful to have in a common base class?  That is, some of these may only apply to numeric fields but StringLimit also derives from this class?

rpastrana replied:
agreed. in fact I must have lost a comment I made on this class. It seems this is prime for an ICfgLimits with maybe 
isValueValid() and perhaps a couple other methods. Implement a numericCfgLimits, and strinCfgLimits, etc.

afishbeck commented on configuration/config2/CfgLimits.hpp:
```diff
@@ -0,0 +1,79 @@
+        int m_minInclusive;
+        int m_maxInclusive;
+        int m_minExclusive;
+        int m_maxExclusive;
```
If there are going to be a huge number of these loaded in memory there are ways you could save some memory.

afishbeck commented on configuration/config2/CfgType.hpp:
```diff
@@ -0,0 +1,57 @@
+        std::shared_ptr<CfgLimits> &getLimits() { return m_pLimits; }
+        void setLimits(const std::shared_ptr<CfgLimits> &pLimits) { m_pLimits = pLimits; }
+        bool isComplete() const { return m_pLimits!=nullptr; }
+        const std::string &getName() { return m_name; }
```
You can probably make many more of these functions const.

afishbeck commented on configuration/config2/CfgValue.hpp:
```diff
@@ -0,0 +1,103 @@
+        std::string m_name;
+        std::string m_displayName;
+        std::string m_mirrorFromPath;
+        bool m_required;
```
These booleans may be more efficiently stored in a bit map.

afishbeck commented on configuration/config2/CfgValue.hpp:
```diff
@@ -0,0 +1,103 @@
+        std::vector<std::shared_ptr<CfgValue>> m_mirrorToCfgValues;
+        std::string m_name;
+        std::string m_displayName;
+        std::string m_mirrorFromPath;
```
if many of these strings are not always set, may make sense to use a map to save memory.

afishbeck commented on configuration/config2/XSDComponentParser.cpp:
```diff
@@ -0,0 +1,155 @@
+namespace pt = boost::property_tree;
+
+
+void XSDComponentParser::parseXSD(const pt::ptree &compTree)
```
more like load than parse once you have a ptree.

afishbeck reviewed (commented):
some comments, but I think it looks mostly ok.  Once we see it in action it will be much clearer what might need to change.

Commit:
Second pass review comments

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 1f980161a9fea4cdfc6be426dbecc0bd8058364c
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |119 |119 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |734 |734 |0 |
|test (thor) |647 |647 |0 |
|test (roxie) |760 |760 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

kenrowland commented:
@rpastrana @afishbeck Please review changes based on last review.

rpastrana commented on configuration/config2/CfgLimits.hpp:
```diff
@@ -50,9 +53,12 @@ class CfgLimits
+        std::vector<AllowedValue> getAllowedValues() const;
+        bool isEnumerated() const { return !m_allowedValues.empty();  }
+        virtual bool isValueValid(const std::string &testValue) { return true; }
+		virtual int getMin() const { return m_minInclusive; }
```
this still looks off

rpastrana commented on configuration/config2/CfgStringLimits.cpp:
```diff
@@ -20,28 +20,7 @@
-
-    }
-
+    std::string limitStr = "";  // this will turn into an error string describing why a value is invalid
```
I'm having a hard time understanding the usefulness of this function, is it meant to be overwritten?
also, if the logic is complete, could we just return ""; ?

rpastrana commented on configuration/config2/EnvironmentMgr.cpp:
```diff
@@ -41,84 +41,69 @@ EnvironmentMgr::EnvironmentMgr(const std::string &configPath) :
-        m_pConfig->postProcessConfig();
+	if (m_pConfigParser)
+	{
+        rc = m_pConfigParser->parseEnvironmentConfig(cfgParms, status);
```
indent looks off in this block

rpastrana commented on configuration/config2/XMLEnvironmentMgr.cpp:
```diff
@@ -84,22 +84,16 @@ void XMLEnvironmentMgr::parse(const pt::ptree &envTree, const std::shared_ptr<Co
-				std::shared_ptr<CfgValue> pCfgValue = pConfigItem->getAttribute(attrIt->first);
-				std::string curValue = attrIt->second.get_value<std::string>();
-				std::shared_ptr<EnvValue> pEnvValue = std::make_shared<EnvValue>(pEnvNode, pCfgValue, attrIt->first, curValue);   // this is where we would use a variant
+              std::shared_ptr<CfgValue> pCfgValue = pConfigItem->getAttribute(attrIt->first);
```
something seems off, there might still be some tabs in this file.

rpastrana commented on configuration/config2/ConfigItem.hpp:
```diff
@@ -0,0 +1,148 @@
+        std::map<std::string, SetInfo> m_uniqueAttributeValueSetDefs;
+
+        // These are the attribute value sets whose members must be unique
+        static std::map<std::string, std::vector<std::shared_ptr<CfgValue>>> m_uniqueAttributeValueSets;    
```
are std:maps thread safe? if not does this static map need to be thread access controlled?

rpastrana commented on configuration/config2/ConfigItem.hpp:
```diff
@@ -0,0 +1,148 @@
+        // These are the attribute value sets whose members must be unique
+        static std::map<std::string, std::vector<std::shared_ptr<CfgValue>>> m_uniqueAttributeValueSets;    
+
+    private:
```
we don't need 139-144, 146

rpastrana reviewed (changes requested):
@kenrowland latest changes seem ok, however most of the original concerns were not addressed.

rpastrana commented on configuration/config2/CfgIntegerLimits.cpp:
```diff
@@ -1,6 +1,6 @@
 /*##############################################################################
 
-HPCC SYSTEMS software Copyright (C) 2017 HPCC Systems�.
+HPCC SYSTEMS software Copyright (C) 2017 HPCC Systems�.
```
github is showing an unknown char immediately after "HPCC Systems"

rpastrana commented on configuration/config2/CfgStringLimits.cpp:
```diff
@@ -64,12 +43,7 @@ bool CfgStringLimits::isValueValid(const std::string &testValue) const
     return isValid;
-
-}
+}
```
new line is needed

JamesDeFabia commented on configuration/config2/EnvValue.cpp:
```diff
@@ -1,6 +1,6 @@
 /*##############################################################################
 
-HPCC SYSTEMS software Copyright (C) 2017 HPCC Systems�.
+HPCC SYSTEMS software Copyright (C) 2017 HPCC Systems.
```
The symbol that is needed after HPCC Systems is the marca registrada or registered trademark symbol ®. 
You can type it using ALT+0174 on numeric keypad in Windows.
If this symbol is problematic, the plain text version (r) or (R) is equivalent.

Commit:
HPCC-17851 Phase 1 core library for new configuration manager

Add core library to support full XSD compliant schema definitions.
Remove custom code in favor of data driven limits and dependencies.

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

Commit:
Merge branch 'config2-phase1' of https://github.com/kenrowland/HPCC-Platform into config2-phase1

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: f44ee4793663c2e2709279cedc81a9a0fc175585
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |92 |92 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |735 |735 |0 |
|test (thor) |648 |648 |0 |
|test (roxie) |764 |764 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   20 sec (00:00:20) |  164 sec (00:02:44) |   58 sec (00:00:58) |    6 sec (00:00:06) |   53 sec (00:00:53) | 1068 sec (00:17:48) |   40 sec (00:00:40) | 1409 sec (00:23:29) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 14ae14d71a0a6ed5878b703d5650f3e45db29c8c
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |92 |92 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |735 |735 |0 |
|test (thor) |648 |648 |0 |
|test (roxie) |764 |764 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   44 sec (00:00:44) |  165 sec (00:02:45) |   59 sec (00:00:59) |    6 sec (00:00:06) |   53 sec (00:00:53) | 1104 sec (00:18:24) |   39 sec (00:00:39) | 1470 sec (00:24:30) |

Commit:
White space changes

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 8a1b471e976f9b20051ea58247359df86af64c92
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |81 |81 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |737 |737 |0 |
|test (thor) |648 |648 |0 |
|test (roxie) |766 |766 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   17 sec (00:00:17) |  165 sec (00:02:45) |   59 sec (00:00:59) |    6 sec (00:00:06) |   53 sec (00:00:53) | 1091 sec (00:18:11) |   39 sec (00:00:39) | 1430 sec (00:23:50) |

kenrowland commented:
Closed in favor of https://github.com/hpcc-systems/HPCC-Platform/pull/10773
